components/sup/src/ctl_gateway/server.rs

```diff
@@ -277,127 +392,20 @@ impl Future for SrvHandler {
-                                                    })
-                                }
-                                "SvcStop" => {
-                                    let m = msg.parse::<protocol::ctl::SvcStop>()
-                                               .map_err(HandlerError::from)?;
-                                    CtlCommand::new(Some(self.tx.clone()),
-                                                    msg.transaction(),
-                                                    move |state, req| {
-                                                        commands::service_stop(state,
-                                                                               req,
-                                                                               m.clone())
-                                                    })
-                                }
-                                "SvcStatus" => {
-                                    let m = msg.parse::<protocol::ctl::SvcStatus>()
-                                               .map_err(HandlerError::from)?;
-                                    CtlCommand::new(Some(self.tx.clone()),
-                                                    msg.transaction(),
-                                                    move |state, req| {
-                                                        commands::service_status(state,
-                                                                                 req,
-                                                                                 m.clone())
-                                                    })
-                                }
-                                "SupDepart" => {
-                                    let m = msg.parse::<protocol::ctl::SupDepart>()
-                                               .map_err(HandlerError::from)?;
-                                    CtlCommand::new(Some(self.tx.clone()),
-                                                    msg.transaction(),
-                                                    move |state, req| {
-                                                        commands::supervisor_depart(state,
-                                                                                    req,
-                                                                                    m.clone())
-                                                    })
-                                }
-                                _ => {
-                                    warn!("Unhandled message, {}", msg.message_id());
-                                    break;
-                                }
-                            };
-                            match self.mgr_tx.start_send(cmd) {
+
+                            let cmd =
+                                match Self::command_from_message(&msg, self.ctl_sender.clone()) {
```

Comment:
This is 3 nested matches inside a loop. Maybe worth breaking into some more functions?

Reply:
Have to think on that a bit. On first glance, it looks like those functions' return values would have to contend with actual Ok/Err values, as well as signaling whether to break or continue from the main loop they'd be embedded in. I'm not seeing an obvious refactoring that wouldn't be more complicated as a result. There may be an alternative way of structuring this entire function to make it more clear, but that may be a bigger (and more risky) job at the moment.